### ISSUE TITLE
MSTK partitioning bug?

### ISSUE THREAD

**Issue by jd-moulton** [amanzi/amanzi#408]
@ecoon reported a possible bug in MSTK:

----

Did Amanzi’s MSTK change somewhat recently?
    
I’m getting some weird errors (still diagnosing) with some old runs and meshes that I’m pretty sure used to work.
    
    In a parallel run with a prepartitioned mesh of type “NFACED,” I see:
    
    FATAL ERROR!! in MESH_ReadExodusII_Serial: Could not find element in sideset
    
on a few (but not all?) processes.  These meshes _do_ have a bunch of sidesets in them (NLCD labels, 11 in total), including some of them that are “sparse” in that they may have only 5-10 out of 50K surface cells in a sideset.  Recall that sidesets are stored as a list of element IDs and a integer side index in that element.
    
The error seems to be complaining that somehow, likely in partitioning (the serial run of this works fine) the element IDs didn’t go with their partition, or things got renumbered inconsistently, or something else, because the process thinks it ought to have a sideset on element X where X is a global ID that is NOT on this process.
    
Is there anything that changed with respect to this recently?  I’m fairly sure these meshes ran in parallel for ExaSheds about 6 months ago.

**Comment by jd-moulton** [comment]
Followup e-mail from @ecoon provided a test

---

Even simpler – here is a 2D transect hillslope with 5 cells in the horizontal and 25 in the vertical.
 
Partitioning this into 2 parts via
 
Mpirun -n 2 meshconvert –classify=1 –partition=1 –partition-method=2 hillslope.exo hillslope.par
(AUTOCORRECT MANGLING!)
 
Should split it into 3 columns in one file and 2 in another.  We should expect ss1 to have 2 or 3 elements in each file.
 
Partition 0 got 2 columns, but has 3 entries:
 
elem_ss1 = 25, 50, 0 ;
side_ss1 = 6, 6, -917546285 ;
 
(the third is junk)
 
Partition 1 got 3 columns, but has 4 entries:
elem_ss1 = 25, 50, 75, 110 ;
side_ss1 = 6, 6, 6, 2043 ;
 
(the fourth is junk).
 
Is it possible this is a ghost cell issue?

**Comment by lipnikov** [comment]
MSTK 3.3.1 Dec 29 ,2019
MSTK 3.2.1 Nov 24, 2019

The error comes from MSTK library.

Today MSTK's commit is "fixed bug in writing out sidesets for parallel files". Looks related.

**Comment by jd-moulton** [comment]
Yes, it’s the fix I’m testing now. I just didn’t want to commit the update to the TPLs without a reference to the bug and tests.
